Add AQL function requests

stephanwilliams opened the pull request:
Adds support for /_api/aqlfunction requests for manipulating AQL user functions (https://docs.arangodb.com/3.2/HTTP/AqlUserFunctions/).

GET /_api/aqlfunction for some reason returns an array instead of the usual wrapper object, so I added a check in senders.go to prevent it from trying to unmarshal the raw response when it's an array (`raw[0] != '['` seems like a bit of a hack, but I don't feel like it will cause any problems in practice).

solher commented:
First of all, thanks a lot for all your work! I'm less using ArangoDB these days and it's really great when people contribute with the requests they use. That's actually why I tried to make it easy to add new requests in the first place :)

solher commented:
About this one, the array check certainly feels hacky but let's keep it until we encounter this issue more frequently. Just one thing, could you add a comment line on it precising that we did it to support the AQL user functions?

stephanwilliams commented:
No problem! This library has been a great help to me so it's nice to be able to contribute back to it.

I came up with some ways of checking the response that are probably more robust:

1. With `json.Decoder` we can parse juuust enough of `raw` to determine whether it's an object before calling `json.Unmarshal`:

```go
func isJSONObject(data []byte) (bool, error) {
	r := bytes.NewReader(data)
	d := json.NewDecoder(r)

	if !d.More() {
	 	 return false, nil
	}

	tok, err := d.Token()
	if err != nil {
		return false, errors.WithStack(err)
	}

	return tok == json.Delim('{'), nil
}
```

2. Or we can check the error when `json.Unmarshal` fails - the `ute.Struct == "" && ute.Field == ""` condition ensures that only errors unmarshaling the top-level object are ignored, and not errors parsing e.g. struct fields.

```go
if err := json.Unmarshal(raw, &parsed); err != nil {
	if ute, ok := err.(*json.UnmarshalTypeError); ok && ute.Struct == "" && ute.Field == "" {
		// Some API calls such as /_api/aqlfunction return arrays instead of objects, so
		// we ignore any *top-level* type errors.
	} else {
		return nil, errors.Wrap(err, "could not decode the json database response")
	}
}
```

Either of these look more appealing?

solher commented:
I'm not totally sure to understand the first solution. What is the difference between this or simply check if the first byte is `{`?

I don't particularly like the aesthetic of the second solution (looks a bit exotic :)) but it is definitely interesting.

Maybe we could just check if the first byte is `{`.

stephanwilliams commented:
Theoretically, the first solution would work in the case where there was leading whitespace in the response, but I don't suppose that that's something we really need to worry about ;)

Changed to `raw[0] == '{'` and added a comment.

solher commented:
LGTM! And thanks to https://github.com/solher/arangolite/pull/31 you'll be able to get the result directly from the Run() method!
